### PR TITLE
IGNITE-10654 Print warn message in case of index creating with already existing fields.

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 1]
```diff
@@ -822,6 +841,70 @@ public boolean rebuildFromHashInProgress() {
...
+     * @return {@code True} if equal index exist.
+     * @throws IgniteCheckedException If failed.
+     */
+    private @Nullable Index checkIndexPresence(Index curIdx) throws IgniteCheckedException {
+        IndexColumn[] curColumns = curIdx.getIndexColumns();
+
+        Index registredIdx = null;
+
+        for (Index idx : idxs) {
+            if (!(idx instanceof H2TreeIndex))
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: when node not in affinity this check is not valid, thus this check correct, isn`t it ?

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 2]
```diff
@@ -822,6 +841,70 @@ public boolean rebuildFromHashInProgress() {
...
+                continue;
+
+            if (F.eq(curIdx.getName(), idx.getName()))
+                throw new IgniteCheckedException("Index already exists: " + idx.getName());
+
+            IndexColumn[] idxColumns = idx.getIndexColumns();
+
+            if (idxColumns.length == curColumns.length) {
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: how idx with _KEY could be created? create index \"idx6\" on Val(valLong, _KEY)" throw :+1: 
javax.cache.CacheException: Column doesn't exist: _KEY

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 3]
```diff
@@ -822,6 +841,70 @@ public boolean rebuildFromHashInProgress() {
...
+                boolean copy = true;
+
+                for (IndexColumn idxCol : idxColumns) {
+                    registredIdx = null;
+
+                    for (IndexColumn curCol : curColumns) {
+                        if (F.eq(aliasNameTransformer(idxCol), aliasNameTransformer(curCol))
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: no, there is no goal to detect such cases here. This fix only for idx with equal fields and search direction.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zstan) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: client check added.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 4]
```diff
@@ -822,6 +841,70 @@ public boolean rebuildFromHashInProgress() {
...
+
+            IndexColumn[] idxColumns = idx.getIndexColumns();
+
+            if (idxColumns.length == curColumns.length) {
+                boolean copy = true;
+
+                for (IndexColumn idxCol : idxColumns) {
+                    registredIdx = null;
+
+                    for (IndexColumn curCol : curColumns) {
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: i have no idea what test i need to add for "unwrapKeyColumns" case, can u suggest?

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 5]
```diff
@@ -822,6 +839,73 @@ public boolean rebuildFromHashInProgress() {
...
+                if (!F.eq(aliasNameTransformer(idxColumns[0]), aliasNameTransformer(curColumns[0])))
+                    continue;
+
+                boolean copy = true;
+
+                for (IndexColumn idxCol : idxColumns) {
+                    registredIdx = null;
+
+                    for (IndexColumn curCol : curColumns) {
+                        if (F.eq(aliasNameTransformer(idxCol), aliasNameTransformer(curCol))
```
Comment (ygerzhedovich) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: seems algorithm of checking indexes is not correct. Need to check columns by their position at index.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 6]
```diff
@@ -822,6 +839,73 @@ public boolean rebuildFromHashInProgress() {
...
+        if (desc.isKeyAliasColumn(col.column.getColumnId()))
+            return QueryUtils.KEY_FIELD_NAME;
+        else if (desc.isValueAliasColumn(col.column.getColumnId()))
+            return QueryUtils.VAL_FIELD_NAME;
+
+        return col.columnName;
+    }
+
+    /**
+     * Check index presence, throw exception if index with same name already exist or return {@code True} if
```
Comment (ygerzhedovich) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: javadoc is incorrect - return value is not boolean

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 7]
```diff
@@ -70,6 +97,15 @@
 
         igniteCfg.setConsistentId(igniteInstanceName);
 
+        if (igniteInstanceName.startsWith(CLIENT_NAME)) {
+            igniteCfg.setClientMode(true);
```
Comment (ygerzhedovich) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: pls, use Ignite format of code

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 8]
```diff
@@ -822,6 +840,67 @@ public boolean rebuildFromHashInProgress() {
...
+
+            for (int i = 0; i < Math.min(idxColumns.length, curColumns.length); ++i) {
+                IndexColumn idxCol = idxColumns[i];
+                IndexColumn curCol = curColumns[i];
+
+                // pk attach at the end of listed fields.
+                if (curCol.column.getColumnId() == 0 && registredIdx != null)
+                    continue;
+
+                if (idxCol.column.getColumnId() == curCol.column.getColumnId() && idxCol.sortType == curCol.sortType)
```
Comment (ygerzhedovich) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: There is H2Utils.equals(IndexColumn c1, IndexColumn c2) method

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java (hunk 6 above)
Comment (ygerzhedovich) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: need to amend javadoc in part 'or return {@code True} if'

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 9]
```diff
@@ -857,6 +873,67 @@ public boolean rebuildFromHashInProgress() {
         return commitUserIndex(ses, idxName);
     }
 
+    /**
+     * Checks and replace if provided column id matches value alias column.
```
Comment (agura) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: Please replace `replace` to `replaces`. Also describe what exactly will be checked and replaced.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 10]
```diff
@@ -857,6 +873,67 @@ public boolean rebuildFromHashInProgress() {
...
+     *
+     * @param col Input column.
+     * @return Transformed result if needed.
+     */
+    private String aliasNameTransformer(IndexColumn col) {
```
Comment (agura) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: Method name is poor. It seems that name should be or should contain some verb that describe action.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 11]
```diff
@@ -857,6 +873,67 @@ public boolean rebuildFromHashInProgress() {
...
+        if (desc.isKeyAliasColumn(col.column.getColumnId()))
+            return QueryUtils.KEY_FIELD_NAME;
+        else if (desc.isValueAliasColumn(col.column.getColumnId()))
+            return QueryUtils.VAL_FIELD_NAME;
+
+        return col.columnName;
+    }
+
+    /**
+     * Check index presence, return {@link Index} if index with same name or same fields and search direction already
```
Comment (agura) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java]: Please replace `Check` to `Checks`.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 12]
```diff
@@ -59,6 +69,24 @@
     /** */
     private int gridCount = 1;
 
+    /** */
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: Please add meaningful comments here and bellow,

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 13]
```diff
@@ -59,6 +69,24 @@
...
+    private ListeningTestLogger srvLog;
+
+    /** */
+    private ListeningTestLogger clntLog;
+
+    /** */
+    private static final String CLIENT_NAME = "client";
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: static final fields should be placed before instance fields.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 14]
```diff
@@ -59,6 +69,24 @@
...
+
+    /** */
+    private static boolean createIdx = true;
+
+    /** */
+    private static boolean createStaticCache = true;
+
+    /** */
+    private static final String TEST_TBL_NAME = "PUBLIC.TEST_TABLE";
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: static final fields should be placed before instance fields.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 15]
```diff
@@ -70,6 +98,17 @@
 
         igniteCfg.setConsistentId(igniteInstanceName);
 
+        if (igniteInstanceName.startsWith(CLIENT_NAME)) {
+            igniteCfg.setClientMode(true);
+
+            if (clntLog != null)
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: What does `clnt` mean? There is no such abbreviation in abbreviation rules (https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules).

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 16]
```diff
@@ -209,6 +256,427 @@ public void testDynamicIndexesNoPersistence() throws Exception {
         }
     }
 
+    /**
+     * Test mixed dynamic and static caches with indexes creation.
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: Test**s**. Here and below.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 17]
```diff
@@ -209,6 +256,427 @@ public void testDynamicIndexesNoPersistence() throws Exception {
...
+
+        populateCache();
+
+        createStaticCache = true;
+
+        try {
+            startGrid(1);
+
+            fail("Exception wasn't thrown");
+        } catch (IgniteCheckedException e) {
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: `catch` must be on the next line.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 18]
```diff
@@ -209,6 +256,427 @@ public void testDynamicIndexesNoPersistence() throws Exception {
...
+
+            sqlIns += reqFlds[i] + ((i < reqFlds.length - 1) ? ", " : ") values (");
+        }
+
+        if (consPkFldsNum > 0) {
+            sql += " CONSTRAINT PK_PERSON PRIMARY KEY (";
+
+            for (int i = 0; i < consPkFldsNum; ++i)
+                sql += reqFlds[i] + ((i < consPkFldsNum - 1) ? ", " : "))");
+        } else
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: `else` must be on the next line.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 19]
```diff
@@ -209,6 +256,427 @@ public void testDynamicIndexesNoPersistence() throws Exception {
...
+            String s0 = sqlIns;
+
+            for (int f = 0; f < reqFlds.length; ++f)
+                s0 += i + ((f < reqFlds.length - 1) ? ", " : ")");
+
+            qryProc.querySqlFields(new SqlFieldsQuery(s0), true).getAll();
+        }
+    }
+
+    /** */
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: Please fix javadoc

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java [hunk 20]
```diff
@@ -209,6 +256,427 @@ public void testDynamicIndexesNoPersistence() throws Exception {
...
+            "LAST_NAME", "LANG", "ADDRESS"));
+
+        assertTrue(checkIdxUsed(qryProc, null, TEST_TBL_NAME, "FIRST_NAME",
+            "LAST_NAME", "ADDRESS", "LANG"));
+        // first idx fields not belongs to request fields.
+        assertTrue(checkIdxUsed(qryProc, "idx2", TEST_TBL_NAME, "ADDRESS",
+            "LAST_NAME"));
+    }
+
+    /** */
```
Comment (agura) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/BasicIndexTest.java]: Please fix javadoc here and below.

**Comment by dspavlov** [comment]
The ticket is resolved, closing PR. 
 
Please close PRs for your resolved tickets and other non-needed PRs. 
 
Alternatively, ask a committer 
- to use GitHub UI (committer can associate accounts here: https://gitbox.apache.org/setup/ )
- or /scripts/apply_pull_request.sh from Apache Ignite source.
- or mention in commit message 'Fixes #(a number of PR)'. 
 
These methods will automatically close PR.
 
Feel free to reopen this PR if it is actual.

**Pull request closed without merging**